frontend/simple/js/transactions.js

```diff
@@ -0,0 +1,263 @@
+  }
+  get isRecoverable () {
+    return _isRecoverable.get(this)
+  }
+  wait () {
+    let wait = _wait.get(this)
+    return wait.promise
+  }
+  addStep (step) {
+    let steps = _steps.get(this)
+    steps.push(step)
+  }
+  // This function creates variables in the transaction's scope to be referenced by steps
+  setInScope (name, value) {
+    let scope = _scope.get(this)
+    scope[name] = value
+    _scope.set(this, scope)
+  }
+  async run (serviceRegistry) {
+    let steps = _steps.get(this)
+    let scope = _scope.get(this)
+    let cursor = _cursor.get(this)
+    let wait = _wait.get(this)
+    for (let i = cursor; i < steps.length; i++) {
+      let args = {}
+      // Keep track of a cursor of what step in the transaction we are operating
+      _cursor.set(this, i)
+      let currentStep = steps[ i ]
+      // Build arguments from services and scoped data
+      for (let argPair of Object.entries(currentStep.args)) {
+        args[argPair[0]] = scope[argPair[1]]
+      }
+      args.setInScope = this.setInScope.bind(this)
+      // Attempt to run step
+      try {
+        await sbp(currentStep.execute, args)
+      } catch (ex) {
+        wait.reject(new Error(`Step - ${currentStep.description || ''} failed with error, \n ${ex}`))
+        throw ex
+      }
+    }
+    wait.resolve()
+  }
+  toJSON () {
+    return {
+      steps: _steps.get(this),
+      scope: _scope.get(this),
+      cursor: _cursor.get(this),
+      isRecoverable: _isRecoverable.get(this),
+      description: _description.get(this)
+    }
+  }
+  static isTransaction (obj) {
+    if (typeof obj !== 'object') return false
+    if (!Array.isArray(obj.steps)) return false
+    if (typeof obj.scope !== 'object') return false
+    if (!Number.isInteger(obj.cursor)) return false
+    if (typeof obj.isRecoverable !== 'boolean') return false
+    if (typeof obj.description !== 'string') return false
+    return true
+  }
+  static fromJSON (json) {
+    if (!Transaction.isTransaction(json)) {
+      throw new TypeError('Invalid Transaction')
+    }
+    for (let key of Object.keys(json.scope)) {
+      if (Events.HashableEntry.isHashableEntry(json.scope[key])) {
+        json.scope[key] = Events.HashableEntry.fromJSON(json.scope[key])
+      }
+    }
+    let transaction = new Transaction(json.description, json.isRecoverable)
+    _scope.set(transaction, json.scope)
+    _cursor.set(transaction, json.cursor)
+    _steps.set(transaction, json.steps)
+    return transaction
+  }
+}
+
+export function createInternalStateTransaction (description) {
+  return new Transaction(description, false)
+}
+
+export function createExternalStateTransaction (description) {
+  return new Transaction(description, true)
+}
+// Transaction Queue
+let _transactionQueue = new WeakMap()
+let _serviceRegistry = new WeakMap()
+let _isRunning = new WeakMap()
+let _transactionID = new WeakMap()
+let _isLoaded = new WeakMap()
+let _failures = new WeakMap()
```

Comment:
Pretty sure these should be written as standard instance properties.

`note-to-self:` consider whether it's simpler to write this file without `TransactionQueue` and `Transaction` classes